modules/renter/proto/contractset.go

```diff
@@ -68,56 +75,70 @@ func (cs *ContractSet) Insert(contract modules.RenterContract) {
 	cs.contracts[contract.ID] = &safeContract{RenterContract: contract}
 }
 
-// Acquire looks up the contract with the specified FileContractID and locks
-// it before returning it. If the contract is not present in the set, Acquire
-// returns false and a zero-valued RenterContract.
-func (cs *ContractSet) Acquire(id types.FileContractID) (modules.RenterContract, bool) {
+// Len returns the number of contracts in the set.
+func (cs *ContractSet) Len() int {
 	cs.mu.Lock()
-	sc, ok := cs.contracts[id]
-	cs.mu.Unlock()
-	if ok {
-		sc.mu.Lock()
-	}
-	return sc.RenterContract, ok
+	defer cs.mu.Unlock()
+	return len(cs.contracts)
 }
 
 // Return returns a locked contract to the set and unlocks it. The contract
 // must have been previously acquired by Acquire. If the contract is not
 // present in the set, Return panics.
 func (cs *ContractSet) Return(contract modules.RenterContract) {
 	cs.mu.Lock()
-	defer cs.mu.Unlock()
-	sc, ok := cs.contracts[contract.ID]
+	safeContract, ok := cs.contracts[contract.ID]
 	if !ok {
+		cs.mu.Unlock()
 		build.Critical("no contract with that id")
 	}
-	sc.RenterContract = contract
-	cs.contracts[contract.ID] = sc
-	sc.mu.Unlock()
+	safeContract.RenterContract = contract
+	cs.contracts[contract.ID] = safeContract
+	cs.mu.Unlock()
+
+	safeContract.mu.Unlock()
 }
 
-// Delete removes a contract from the set. The contract must have been
-// previously acquired by Acquire. If the contract is not present in the set,
-// Delete is a no-op.
-func (cs *ContractSet) Delete(contract modules.RenterContract) {
+// View returns a copy of the contract with the specified ID. The contracts is
+// not locked. Certain fields, including the MerkleRoots, are set to nil for
+// safety reasons. If the contract is not present in the set, View
+// returns false and a zero-valued RenterContract.
+func (cs *ContractSet) View(id types.FileContractID) (modules.RenterContract, bool) {
 	cs.mu.Lock()
 	defer cs.mu.Unlock()
-	sc, ok := cs.contracts[contract.ID]
+	safeContract, ok := cs.contracts[id]
 	if !ok {
-		return
+		return modules.RenterContract{}, false
 	}
-	delete(cs.contracts, contract.ID)
-	sc.mu.Unlock()
+	c := safeContract.RenterContract
+	c.MerkleRoots = nil
+	return c, true
+}
+
+// ViewAll returns a copy of each contract in the set. The contracts are not
+// locked. Certain fields, including the MerkleRoots, are set to nil for
+// safety reasons.
```

Comment:
Can you elaborate on what are the reasons to reset MerkleRoots, please?

Reply:
You should only be able to modify the merkle roots if you hold the contract's lock. `ViewAll` provides the `modules.RenterContract` but doesn't acquire the contract's lock. It's okay to return the `modules.RenterContract` because it's copied by value, so callers can't modify shared data. But `MerkleRoots` is a slice, and copying the slice pointer still gives you access to the underlying (shared) data. So we either have to copy the whole slice (not practical here, since the Merkle roots may be large), or set it to nil.